### PR TITLE
Enhance filesystem dock tooltips

### LINKED ISSUE #175: Show images in a tooltip when hovering them in the FileSystem dock
**Describe the project you are working on:**
I'm currently running through the docs to learn how to use Godot.  I'm planning on doing some Game Jams with it, and resurrecting some older projects.

**Describe the problem or limitation you are having in your project:**
Not a limitation, but a usability enhancement for Godot.  Right now when trying to identify images in the `FileSystem` panel, all you can rely on is the name of the file, or the tiny preview icon.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Right now when mousing over an image in the `FileSystem` panel, it will only show the filename in the tooltip.  I think it would be more useful if the actual image (or a preview of it) showed up in the tooltip.  I've seen this done in all sorts of applications.  

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
I think this is simple enough to not need a mockup creation, but I can provide one if requested.

**Describe implementation detail for your proposal (in code), if possible:**
Pseudocode provided by @Jummit:
```
if file is Image then
   set_custom_tooltip(new ImageTexture().from_data(file.data))
else
   set_tooltip(file.name)
end
```

I do think that there should be a hard limit on the tooltip size.  e.g. 256x256 pixels max.  Anything larger should be scaled down.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
N/A

**Is there a reason why this should be core and not an add-on in the asset library?:**
Very nice (and hopefully simple) usability enhancement.

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/175

Added this to the tooltips:
- file size
- resource type (if resource)
- bigger preview (if texture or image)
godot windows tools 64_BQPpwf9LeL

### CODE REVIEW

**Comment by YeldhamDev** [comment]
The borders of the tooltips for images seem broken.

**Comment by KoBeWi** [comment]
What do you mean?
The bigger tooltip in the GIF just went outside recording area 🙃

**Comment by YeldhamDev** [comment]
> What do you mean?

The left border has a much smaller gap than the rest when hovering images.

**Comment by KoBeWi** [comment]
Borders are managed by the default tooltip container, so it doesn't seem like my fault. I could maybe hack it by adding an empty control.

**Comment by YeldhamDev** [comment]
It would be better if we found the root issue rather than implementing a hack. Still, not a stopper for this PR.

**Comment by fire-forge** [comment]
What do you think about adding previews for more resources there, using the same system the FileSystem uses to get previews in grid mode? That way more resources, like scenes, scripts, gradients, materials, fonts, etc. could have tooltip previews.

image

**Comment by KoBeWi** [comment]
I'd leave it for later. The PR lays a nice foundation for such additions.

File: editor/filesystem_dock.cpp
Comment (fire-forge) [inline editor/filesystem_dock.cpp:2175]:
```suggestion
				Label *label = memnew(Label(TTR("Dimensions:") + vformat(String::utf8(" %s×%s"), preview->get_width(), preview->get_height())));
```

See https://github.com/godotengine/godot/pull/60008#discussion_r847653853

File: editor/filesystem_dock.cpp
Comment (fire-forge) [inline editor/filesystem_dock.cpp:2183]:
```suggestion
				Label *label = memnew(Label(TTR("Dimensions:") + vformat(String::utf8(" %s×%s"), preview->get_width(), preview->get_height())));
```

**Review by fire-forge — APPROVED** [review]

File: editor/filesystem_dock.cpp
Comment (timothyqiu) [reply editor/filesystem_dock.cpp:2175]: I would suggest moving the translation into the vformat call. So that the colon & space after it can be properly localized. e.g. CJK usually prefer removing the space after a full-width colon.

```suggestion
				Label *label = memnew(Label(vformat(TTR("Dimensions: %d×%d"), preview->get_width(), preview->get_height())));
```

Same to other hardcoded whitespace here.

**Comment by groud** [comment]
I don't am not sure the preview implementation is acceptable as loading a resource can be costly. In that case it's only images, but I think I would have preferred that this PR uses the built in resource preview system.

It should even simplify the implementation, as it's fairly easy to use. You can check for uses of `EditorResourcePreview::get_singleton()->queue_resource_preview` to check how it works.

**Comment by KoBeWi** [comment]
I tried this
```C++
TextureRect *tr = memnew(TextureRect);
EditorResourcePreview::get_singleton()->queue_resource_preview(p_path, tr, "set_texture", Variant());
```
but it doesn't have effect 🤔

**Comment by groud** [comment]
Yeah I agree with the fact that `EditorResourcePreview` could also generate some metadata to be displayed along with the picture or something like that.
If you don't want to implement that right now and maybe to avoid freezes, maybe you could load the texture in the `_tooltip_thumbnail_done` callback? With some luck the texture will still be kept loaded and it would avoid the loading freeze.

**Comment by KoBeWi** [comment]
It doesn't work. When the preview is cached it doesn't even load the texture. I can look into implementing preview metadata, but I don't think it really blocks this PR.

**Comment by reduz** [comment]
All welcome the new Godot Operating System overlord.

File: editor/filesystem_dock.cpp
Comment (reduz) [inline editor/filesystem_dock.cpp:2153]: Well, this is kind of hacky and you know how this story continues in Godot. After the PR is merged, you get a proposal opened:

"Nice resource file info tooltip you have there, can I do this for my custom resource type from my extension?" 

So I think this should be a bit more pluggable. So maybe it would be good to make a tiny class: `EditorResourceFileTooltipPlugin` and register them directly into the FileSystem dock? (via EditorPlugin API).

**Comment by mhilbrunner** [comment]
Editor PR review meeting: feature is wanted and approved, @KoBeWi wants to improve texture loading first seperately.

File: editor/filesystem_dock.cpp
Comment (KoBeWi) [reply editor/filesystem_dock.cpp:2153]: That's too much work, I'd rather wait for someone to actually request it (the code won't be difficult to replace).

**Comment by KoBeWi** [comment]
Rebased after #64628. For now this PR still uses the hard-coded approach. I'll look into implementing EditorResourceFileTooltipPlugin.

**Comment by KoBeWi** [comment]
Ok I added an API for creating resource tooltips and added a plugin for textures. We can add more plugins later (I have some interesting idea for audio files).

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
